### ISSUE TITLE
backfiller should verify segment hashes

### ISSUE THREAD

**Issue by ekimekim** [ekimekim/wubloader#80]
When backfilling, the backfiller should ensure the file it recieves matches the hash in the filename.

What to do when you have a mismatch is an open question. Let's think about what kinds of failures we expect to detect:
1. Corruption in transit - this is probably unlikely, and when it does happen it will almost always be due to the connection being closed early and the file truncated. In these cases it makes the most sense to just retry.
2. Corruption on disk on remote end - this is definitely a possibility, assuming we're not doing anything fancy like ZFS. In this case retrying won't help, and our only real remediation would be to delete the segment and let it replicate from another node. Of course, if that's the only copy, we have more issues. This should be a manual intervention, so the code should just error so we notice.
3. The file is intact but the hash is wrong - This is extremely unlikely. It would require either a particularly egregious logic error in the downloader, or a memory corruption bug (or disk corruption in the inode, but at that point you're probably going to corrupt the whole filesystem). Short of buying all nodes ECC RAM, there's not much we can do about it. This could also happen if someone is manually editing the segment filenames.

Basically, corruption in transit is the main thing to worry about and should be automatically retried. Beyond that, problems are probably hardware-related and require manual intervention. So I suggest we emit a warning, discard the recieved file, and just let it be retried next time we backfill. We should also add metrics for this failure condition.

**Comment by chrusher** [comment]
I'll take a crack at this.
